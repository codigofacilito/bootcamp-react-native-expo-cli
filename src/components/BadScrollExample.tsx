import React, { useEffect, useState } from 'react';
import { ScrollView, Text, View } from 'react-native';

export default function BadScrollExample() {
  const [items, setItems] = useState<any[]>([]);
  const [loading, setLoading] = useState(true);
  const [error, setError] = useState<string | null>(null);

  useEffect(() => {
    setError(null);
    setLoading(true);
    fetch('https://jsonplaceholder.typicode.com/posts') // simula una respuesta larga
      .then((res) => res.json())
      .then((data) => {
        // Simulamos tener 3000 elementos duplicando el resultado
        const hugeList = Array(30).fill(data).flat();
        setItems(hugeList);
        setLoading(false);
      })
      .catch((err) => {
        console.error('Error fetching items:', err);
        setError('Error loading items');
        setLoading(false);
      });
    // Simula una API que devuelve 1000+ elementos
  }, []);

  // Loading es mala practica de UX
  if (loading) {
    return (
      <View style={{ flex: 1, justifyContent: 'center', alignItems: 'center' }} testID='loading-indicator'>
        <View>
          <Text>Loading items...</Text>
        </View>
      </View>
    );
  }

  if (error) {
    return (
      <View style={{ flex: 1, justifyContent: 'center', alignItems: 'center' }} testID='error-indicator'>
        <Text>Error loading items</Text>
      </View>
    );
  }

  return (
    <ScrollView>
      {/** Nunca utilizes o renderices directamente sin una Lista */}
      {items.map((item, index) => (
        <View key={`post-item-${item.key}-${index}`/** Key */} style={{ padding: 10, borderBottomWidth: 1 }}>
          <Text style={{ fontWeight: 'bold' }}>{item.title}</Text>
          <Text>{item.body}</Text>
        </View>
      ))}
    </ScrollView>
  );
}
