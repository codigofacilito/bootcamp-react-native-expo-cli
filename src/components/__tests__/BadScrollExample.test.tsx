import { render } from "@testing-library/react-native";
import BadScrollExample from "../BadScrollExample";


describe("<BadScrollExample /> component", () => {
    test("Renderea correctamente", () => {
        const { toJSON, getByTestId } = render(<BadScrollExample />);

        expect(toJSON()).toMatchSnapshot();
        expect(getByTestId("activity-indicator")).toBeDefined();
    });

    const fetch = () => {
        return {
            json: () => {
                return {};
                // creamos la funcion para mostrar los 30 elementos
            }
        }
    };
    test("Renderea elementos despues de cargarlos", async () => {
        jest.spyOn(global, "fetch").mockResolvedValue({
            json: jest.fn().mockResolvedValue(Array.from({ length: 30 }, (_, i) => ({
                userId: 1,
                id: i + 1,
                title: `Post ${i + 1}`,
                body: `This is the body of post ${i + 1}.`
            }))),
        } as unknown as Response);

        const { findAllByText } = render(<BadScrollExample />);

        const results = await findAllByText(/post/i);
        expect(results.length).toBeGreaterThan(0);
        expect(results).toBeDefined();
    });
});