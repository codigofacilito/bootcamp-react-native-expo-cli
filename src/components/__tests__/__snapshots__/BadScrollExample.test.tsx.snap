// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<BadScrollExample /> component Renderea correctamente 1`] = `
<View
  style={
    {
      "alignItems": "center",
      "flex": 1,
      "justifyContent": "center",
    }
  }
>
  <ActivityIndicator
    size="large"
    testID="activity-indicator"
  />
</View>
`;
